sycl/include/CL/sycl/detail/context_impl.hpp

```diff
@@ -58,12 +83,33 @@ class context_impl {
   RT::PiContext &getHandleRef();
   const RT::PiContext &getHandleRef() const;
 
-  std::map<OSModuleHandle, RT::PiProgram> &getCachedPrograms() {
+  locked<ProgramCacheT> acquireCachedPrograms() {
+    return {m_CachedPrograms, m_ProgramCacheMutex};
+  }
+
+  // Non thread-safe
+  ProgramCacheT &getCachedPrograms() {
     return m_CachedPrograms;
   }
-  std::map<RT::PiProgram, std::map<string_class, RT::PiKernel>> &
-  getCachedKernels() {
-    return m_CachedKernels;
+
+  locked<KernelCacheT> acquireKernelsPerProgramCache() {
+    return {m_KernelsPerProgramCache, m_KernelsPerProgramCacheMutex};
+  }
+
+  // Non thread-safe
+  KernelCacheT &getKernelsPerProgramCache() {
+    return m_KernelsPerProgramCache;
+  }
+
+  template<class Predicate>
+  void waitProgramBuild(Predicate Pred) {
+    std::unique_lock<std::mutex> Lock(m_ProgramBuildCVMutex);
+
+    m_ProgramBuildCV.wait(Lock, Pred);
+  }
+
+  void notifyAllProgramBuild() {
```

Comment:
It seems context_impl gets a lot of unrelated methods and becomes "one class for everything". Can we move all this program related functionality to a separate entity?